### ISSUE TITLE
parse-string should not accept trailing junk

### ISSUE THREAD

**Issue by cespare** [dakrone/cheshire#65]
Right now this works:

``` clojure
(cheshire.core/parse-string "{\"foo\": 1}asdf") ; => {"foo" 1}
```

Instead, it should raise an exception about the trailing `asdf`. This is confusing and incorrect behavior for a function called `parse-string`, which reads a single JSON value from a string. It only makes sense that the entire string should be a valid JSON object.

(This issue is related to #52 -- in that bug, you identified that a string is a valid JSON value, which is certainly true; however, the problem there, as here, is that `parse-string` accepts trailing junk.)

The reason this bug exists (and the reason the same exact bug is in `data.json` and `clj-json`) is that `parse-string` is implemented using a streaming JSON reader (jackson in this case). The (pseudo-)code is essentially

```
(read-json-object-from-stream (StringReader. s))
```

If you use a streaming parser on a single string like this, you need to also check that the `StringReader` is empty (drained) after parsing the JSON value.

For comparison, Go, Javascript, Ruby, and Python all have JSON parsers in their standard libraries which all have some kind of `parse-string` equivalent function that rejects trailing data.

**Comment by philc** [comment]
Any update on this? Would love to see a fix.

**Comment by Satshabad** [comment]
Also would love to see a fix

**Comment by mkwatson** [comment]
Me too! This was very surprising!

**Comment by overthink** [comment]
+1 Just bit me also.

**Comment by dakrone** [comment]
Closed by #122

**Comment by cespare** [comment]
I still think it's misleading that the default `parse-string` function accepts a string that's invalid JSON, but at least there's a workaround for people after they're bitten by this :|
